rust/kernel/net/tcp/cong.rs

```diff
@@ -0,0 +1,544 @@
+//! Congestion control algorithms (CCA).
+//!
+//! Abstractions for implementing pluggable CCAs in Rust.
+#![allow(dead_code)]
+#![allow(unused_variables)]
+#![allow(missing_docs)]
+
+use crate::bindings;
+use crate::error::{self, Error};
+use crate::init::PinInit;
+use crate::prelude::{pr_err, vtable};
+use crate::str::CStr;
+use crate::types::Opaque;
+use crate::ThisModule;
+use crate::{build_assert, field_size, try_pin_init};
+use core::convert::TryFrom;
+use core::marker::PhantomData;
+use core::pin::Pin;
+use macros::{pin_data, pinned_drop};
+
+use super::{InetConnectionSock, TcpSock};
+
+/// Congestion control algorithm (CCA).
+///
+/// A CCA is implemented as a set of callbacks that are invoked whenever
+/// specific events occur in a connection. Each socket has its own instance of
+/// some CCA. Every instance of a CCA has its own private data that is stored in
+/// the socket and is mutated by the callbacks.
+///
+/// Callbacks that operate on the same instance are guaranteed to run
+/// sequentially, and each callback has exclusive mutable access to the private
+/// data of the instance it operates on.
+#[vtable]
+pub trait Algorithm {
+    /// Private data. Each socket has its own instance.
+    type Data: Default + Send + Sync;
+
+    /// Name of the algorithm.
+    const NAME: &'static CStr;
+
+    /// Called when entering CWR, Recovery, or Loss states from Open or Disorder
+    /// states. Returns the new slow start threshold.
+    fn ssthresh(sk: &mut Sock<Self>) -> u32;
+
+    /// Called towards the end of processing an ACK if a cwnd increase is
+    /// possible. Performs a new cwnd calculation and sets it on the socket.
+    // Note: In fact, one of `cong_avoid` and `cond_control` is required.
+    // (see `tcp_validate_congestion_control`)
+    fn cong_avoid(sk: &mut Sock<Self>, ack: u32, acked: u32);
+
+    /// Called before the sender's congestion state is changed.
+    fn set_state(sk: &mut Sock<Self>, new_state: State) {}
```

Comment:
These default implementations should have a `kernel::build_error(VTABLE_DEFAULT_ERROR)`. My explanation at https://lore.kernel.org/rust-for-linux/CALNs47svLHD3=LFrhi=zf4hdr--e6hGjTzT5X_U9yd8q5r7G7g@mail.gmail.com/ may help